### PR TITLE
Fs multiget

### PR DESCRIPTION
**Pull request by martindurant**
For async IO. 

Read-only for now, and needs fsspec master

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

**Comment by martindurant** [comment]
Thoughts?

File: requirements_dev_optional.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ pytest-cov==2.7.1
 pytest-doctestplus==0.4.0
 pytest-remotedata==0.3.2
 h5py==2.10.0
-s3fs==0.5.0; python_version > '3.6'
+s3fs==0.5.1; python_version > '3.6'
+git+https://github.com/intake/filesystem_spec; python_version > '3.6'
```
Comment (jakirkham) [inline requirements_dev_optional.txt:22]: Is there a package on PyPI or conda-forge that we could install?

Comment (martindurant) [reply requirements_dev_optional.txt:22]: Not yet, no.

Comment (jakirkham) [reply requirements_dev_optional.txt:22]: Do we have a rough sense when these might be created?

Comment (martindurant) [reply requirements_dev_optional.txt:22]: When everything is good here, I can be sure to make a release before merging, so that this line can be reverted

**Comment by alimanfoo** [comment]
We're  just reviewing this now in the zarr v3 dev call. Looks good to me :-)

File: requirements_dev_optional.txt (hunk 1 above)
Comment (alimanfoo) [reply requirements_dev_optional.txt:22]: Sounds good @martindurant.

**Comment by martindurant** [comment]
@alimanfoo , can I get your +1 on this?

**Comment by martindurant** [comment]
@jakirkham @joshmoore , any more thoughts?
I'd like to merge. Since this only affects fsspec storage, which is not used by anyone in the wold yet, I think this is safe.

**Comment by rabernat** [comment]
Just catching up on this PR. Fantastic work @martindurant!

Are we planning to eventually do `multiset` as well?

**Comment by martindurant** [comment]
Yes, of course. I have local code that is partway there.

**Comment by martindurant** [comment]
Yes, the (test) requirement that went in with this PR is for the latest release fsspec. I don't know if we want to say anything about this functionality's dependency in the release notes.

**Comment by jakirkham** [comment]
Good point. If there is some wording you'd like to see in the release notes, would suggest proposing it in PR ( https://github.com/zarr-developers/zarr-python/pull/616 )

**Comment by martindurant** [comment]
Hm, not sure whether to say use fsspec>=0.8.0 or to fallback to sequential

**Comment by cgohlke** [comment]
Is the `getitems` method documented?  Sorry if I missed something obvious, but I could not find it. I try to implement `getitems` for a TIFF store.

**Comment by martindurant** [comment]
https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.FSMap.getitems

**Comment by cgohlke** [comment]
Thank you. I don't use fsspec but try to implement `getitems` independently. My understanding is that the following should work? Somehow it returns garbage
```
    def getitems(self, keys):
        return {key: self.__getitem__(key) for key in keys}
```

**Comment by martindurant** [comment]
That looks OK to me, but that's what the current implementation is effectively doing anyway, with the loop outside of the storage class.

**Comment by jakirkham** [comment]
I think we would still be open to a PR if someone wants to implement that directly in Zarr.

**Comment by martindurant** [comment]
It would may be sensible to make a fallback implementation of getitems in the storage class, doing the same loop currently higher up; but that would require a strict class hierarchy and not actually make anything simpler overall.

**Comment by cgohlke** [comment]
False alarm, `return {key: self.__getitem__(key) for key in keys}` works as expected. I had forgotten to comment out a `yield` statement in the function that was left from previous tries to implement `getitems` as a generator...

**Comment by jakirkham** [comment]
> It would may be sensible to make a fallback implementation of getitems in the storage class, doing the same loop currently higher up; but that would require a strict class hierarchy and not actually make anything simpler overall.

I don't think that is strictly true. We already implement functions that have fallbacks (like [`rmdir`]( https://github.com/zarr-developers/zarr-python/blob/e3cdd1a50e1047304be2c91a017da6362f8df533/zarr/storage.py#L110 )) for storage classes that don't have them. So we could presumably do the same thing with `getitems`. The main benefit would be other functions in `Group` and `Array` could rely on `getitems` instead of `getitem` more generally.

That said, I think having a class hierarchy would be a nice improvement as long as we still have a way to coerce generic `MutableMapping`-based objects into something this hierarchy can work with (perhaps through a wrapper class). This mostly just reiterates what has come up in a few meeting now though.

**Pull request merged**